### PR TITLE
sitemap generation

### PR DESCRIPTION
**Pull request by pnavarrc**
Implement a script to generate the sitemap and integrate it in the build process.

### CODE REVIEW

**Comment by pnavarrc** [comment]
[SEO: Add a site-map to surface metro pages](https://trello.com/c/1khF13CJ/696-seo-add-a-site-map-to-surface-metro-pages)

File: scripts/generate_sitemap/utils.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  const homeMapUrl = urlJoin(shareImageBaseUrl, 'home.png');
+
+  const items = [
+    {
+      url: '/',
+      img: [
+        {
+          url: homeMapUrl,
+          caption: 'Realtime US COVID Risk Map & Key Metrics',
+          title: 'Realtime US COVID Risk Map & Key Metrics',
```
Comment (mikelehen) [inline scripts/generate_sitemap/utils.ts:73]: Should we be unifying around our new homepage title ("U.S. COVID Map & Risk Levels") ?

File: scripts/generate_sitemap/index.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import { concat } from 'lodash';
+import { simpleSitemapAndIndex } from 'sitemap';
+import {
+  getLearnPageItems,
+  getTopLevelPageItems,
+  getLocationPageItems,
+  getHomeImageItems,
+} from './utils';
+
+const destinationDir = path.join(__dirname, '../../public');
```
Comment (mikelehen) [inline scripts/generate_sitemap/index.ts:11]: Are you intending that we check this file in?  Or do we generate it as part of our deploy process?  If the latter, I think it should probably go in `../../build` instead?

File: scripts/generate_sitemap/utils.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    '/terms',
+    '/privacy',
+  ];
+  return relativeUrlList.map(relativeUrl => ({
+    url: relativeUrl,
+  }));
+}
+
+export function getHomeImageItems(): SitemapItemLoose[] {
+  const homeMapUrl = urlJoin(shareImageBaseUrl, 'home.png');
```
Comment (mikelehen) [inline scripts/generate_sitemap/utils.ts:64]: Oooh nice!  Though this is the sharing image but I think we probably want to use the one I made to put in our `<meta name='image'>` tag instead: https://github.com/covid-projections/covid-projections/blob/40c34030f374e2da845a79bd4f06bd6721c53e6e/src/common/urls.ts#L48

File: scripts/generate_sitemap/utils.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  const items = [
+    {
+      url: '/',
+      img: [
+        {
+          url: homeMapUrl,
+          caption: 'Realtime US COVID Risk Map & Key Metrics',
+          title: 'Realtime US COVID Risk Map & Key Metrics',
+          geoLocation: 'United States',
+          // license
```
Comment (mikelehen) [inline scripts/generate_sitemap/utils.ts:75]: what is this? should it just be removed?

File: scripts/generate_sitemap/utils.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+export function getLearnPageItems(): SitemapItemLoose[] {
+  const topLevelRelativeUrls = [
+    '/learn',
+    '/glossary',
+    '/faq',
+    '/case-studies',
+    '/deep-dives',
+    '/covid-risk-levels-metrics',
+  ];
```
Comment (mikelehen) [inline scripts/generate_sitemap/utils.ts:27]: Is there a way we can avoid hard-coding these so it doesn't get out-of date?  Else, is there a way we can create "cross-referenced" comments between here and our router / whatever to help us remember to keep these in sync?  Or somehow we can have a test that catches if things get out of sync?

That said, I guess I don't know for sure what happens if we get out-of-sync.  I assume google will still crawl any URL it finds on our site, so maybe it doesn't matter too much.  In that case, maybe we don't even need to bother including everything in the sitemap?  I'm not sure...  Any idea?

**Review by mikelehen — COMMENTED** [review]
🎉  

Couple suggestions, but looks good!

Seems like we'll want to integrate this `yarn build` so this happens automatically as part of deploys. Is that what you're thinking?

File: scripts/generate_sitemap/utils.ts (hunk 5 above)
Comment (ghop02) [reply scripts/generate_sitemap/utils.ts:27]: This may be overkill, but looks like there's an npm package that helps with it: https://github.com/kuflash/react-router-sitemap

File: scripts/generate_sitemap/index.ts (hunk 2 above)
Comment (pnavarrc) [reply scripts/generate_sitemap/index.ts:11]: I think we need to generate it as part of the deploy process - it needs to be available under `https://covidactnow.org/` for it to work, so I think it needs to go in `../../build`

File: scripts/generate_sitemap/utils.ts (hunk 5 above)
Comment (pnavarrc) [reply scripts/generate_sitemap/utils.ts:27]: I investigated `react-router-sitemap` - there is a simpler way, just export the router config for each family of pages and use it to generate the sitemap. This will help us keep the router and sitemap in sync 👌

File: scripts/generate_sitemap/utils.ts (hunk 4 above)
Comment (pnavarrc) [reply scripts/generate_sitemap/utils.ts:75]: we can add the license, but we don't have to, removing!

**Comment by mikelehen** [comment]
Let me know if/when you want a final review.

**Comment by pnavarrc** [comment]
I'm going to have something cleaner tomorrow morning!

File: scripts/generate_sitemap/utils.ts (hunk 5 above)
Comment (pnavarrc) [reply scripts/generate_sitemap/utils.ts:27]: Actually I looked into this a bit more, in order to centralize the config we will need to load all the possible URL params in the router config, which we probably don't want to do.

File: scripts/generate_sitemap/utils.ts (hunk 4 above)
Comment (pnavarrc) [reply scripts/generate_sitemap/utils.ts:75]: The package doesn't generate exactly the sitemap markup needed by Google, I'm removing the image part for now to avoid delaying the metro areas.

File: scripts/generate_sitemap/index.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+function main() {
+  console.log('Generating the sitemap');
+  const topLevelPageItems = getTopLevelPageItems();
+  const locationPageItems = getLocationPageItems();
+  const learnPageItems = getLearnPageItems();
+
+  const urlList = concat(topLevelPageItems, locationPageItems, learnPageItems);
+  console.log(`Total URLs: ${urlList.length}`);
+
+  simpleSitemapAndIndex({
```
Comment (mikelehen) [inline scripts/generate_sitemap/index.ts:21]: This returns a promise but we're ignoring it, which probably means:
1. We'll log 'Done' before the sitemap is generated.
2. If there are any errors, they won't fail the script.

I think we need to add:

```
.catch(err => {
  console.error(err);
  process.exit(-1);
}).then(() => { 
  console.log('Done.');
});
```

or else rework it to use async/await.  Note that if we use async / await we'll still need a top-level `catch` handler that calls `process.exit`, similar to https://github.com/covid-projections/covid-projections/blob/267ff9ea4155fb17f6eb78012c5b7f95c12283bd/scripts/generate_index_pages.ts#L263

File: src/App.js [hunk 7]
```diff
@@ -152,12 +152,12 @@ export default function App() {
...
-                exact
-                path="/state/:stateId/county/:countyId"
-                component={LocationPage}
-              />
+              <Route exact path="/state/:stateId">
+                <Redirect to="/us/:stateId" />
+              </Route>
+              <Route exact path="/state/:stateId/county/:countyId">
+                <Redirect to="/us/:stateId/county/:countyId" />
+              </Route>
```
Comment (mikelehen) [inline src/App.js:160]: 1. This seems unrelated to sitemap, so just want to confirm this is intentional.
2. Won't this redirect to our "old" URLs?  (i.e. `/us/wa` instead of `/us/washington-wa/`)?  Is that useful?

**Review by mikelehen — APPROVED** [review]

File: src/App.js (hunk 7 above)
Comment (pnavarrc) [reply src/App.js:160]: One way to tell Google which page is the canonical is to redirect to the canonical version. I'm thinking of doing this in 2 steps:

1. redirect `state/wa` → `us/wa`
2. redirect `us/wa` → `us/washington-wa` (on a separate PR)

The second redirect is a bit more tricky, because we need to write the code to find the region by state code and redirect, the first one is straightforward. I can remove this change from this PR and implement both things on a separate PR if that's better!

Comment (pnavarrc) [reply src/App.js:160]: I implemented the redirect on a follow up PR https://github.com/covid-projections/covid-projections/pull/2279
